File: src/once/tools.ts
/** Tofu / Ansible tool workflows. */
import {
  ERR,
  EXIT,
  RENDER_TEMPLATES,
  RUN_SHELL_OPTS,
  WF_NAME,
  WF_OBJECT_FN,
  WF_OBJECT_PREFIX,
  WF_PARAMS,
  WF_PATH_FN,
  WF_PREFIX,
  WF_STEPS,
  type Opts,
  type StepFn,
} from "big-config";
import { Construct, addSuffix, construct } from "big-config/big-tofu/core";
import { registerHandleStep } from "big-config/pluggable";
import { templates as renderTemplates } from "big-config/render";
import { createExitStepFn, createPrintErrorStepFn } from "big-config/step-fns";
import { deepMerge, keywordToPath, sortNestedMap } from "big-config/utils";
import { parseArgs, prepare, printStepFn, runSteps } from "big-config/workflow";

const END = "big-config.workflow/end";

export const stepFns: StepFn[] = [
  printStepFn,
  createExitStepFn(END),
  createPrintErrorStepFn(END),
];

export const delimiters = {
  "tag-open": "<",
  "tag-close": ">",
  "filter-open": "{",
  "filter-close": "}",
};

export const TOFU = "io.github.bigconig-ai.once.tools/tofu";
export const TOFU_SMTP = "io.github.bigconig-ai.once.tools/tofu-smtp";
export const TOFU_DNS = "io.github.bigconig-ai.once.tools/tofu-dns";
export const TOFU_SMTP_POST = "io.github.bigconig-ai.once.tools/tofu-smtp-post";
export const ANSIBLE_LOCAL = "io.github.bigconig-ai.once.tools/ansible-local";
export const ANSIBLE = "io.github.bigconig-ai.once.tools/ansible";

export const pluginStep = "io.github.bigconig-ai.once.tools/render-tofu-backend";

function runStepsWithPlugin(plugin: string, sfns: StepFn[], opts: Opts): Opts {
  const steps: string[] = [];
  for (const step of opts[WF_STEPS] ?? []) {
    if (step === "render") steps.push(step, plugin);
    else steps.push(step);
  }
  return runSteps(sfns, { ...opts, [WF_STEPS]: steps });
}

function providerParam(opts: Opts, key: string, defaultValue: any): any {
  return (opts[WF_PARAMS] ?? {})[key] ?? defaultValue;
}

registerHandleStep(pluginStep, (_f, _step, sfns, opts) => {
  const providerBackend = providerParam(opts, "provider-backend", "s3");
  const prepareKeys = [WF_NAME, WF_PATH_FN, WF_PREFIX, WF_OBJECT_FN, WF_OBJECT_PREFIX, WF_PARAMS];
  const overrides: Opts = {};
  for (const k of prepareKeys) if (k in opts) overrides[k] = opts[k];
  const prepared = prepare(
    {
      [WF_NAME]: opts[WF_NAME],
      [RENDER_TEMPLATES]: [
        {
          template: keywordToPath("io.github.bigconig-ai.once.tools/tofu-backend"),
          overwrite: true,
          "provider-backend": providerBackend,
          transform: [[providerBackend, delimiters]],
        },
      ],
    },
    overrides,
  );
  const pluginOpts = renderTemplates(sfns, prepared);
  return {
    ...opts,
    [EXIT]: pluginOpts[EXIT],
    [ERR]: pluginOpts[ERR],
    [pluginStep]: [...(opts[pluginStep] ?? []), pluginOpts],
  };
});

function ipDataFn(data: Record<string, any>): Record<string, any> {
  return { ...data, ip: data.ip ?? "192.168.0.1" };
}

function cljJson(value: any, indent = 0): string {
  const sp = " ".repeat(indent);
  const child = " ".repeat(indent + 2);
  if (Array.isArray(value)) {
    if (value.length === 0) return "[ ]";
    return [
      "[",
      ...value.map((v, i) => `${child}${cljJson(v, indent + 2)}${i < value.length - 1 ? "," : ""}`),
      `${sp}]`,
    ].join("\n");
  }
  if (value && typeof value === "object") {
    const entries = Object.entries(value);
    if (entries.length === 0) return "{ }";
    return [
      "{",
      ...entries.map(([k, v], i) => `${child}${JSON.stringify(k)} : ${cljJson(v, indent + 2)}${i < entries.length - 1 ? "," : ""}`),
      `${sp}}`,
    ].join("\n");
  }
  if (typeof value === "string") return JSON.stringify(value);
  if (value === true) return "true";
  if (value === false) return "false";
  if (value == null) return "null";
  return String(value);
}

/** Build the Cloudflare DNS records JSON from the SMTP records. */
export function renderFn(src: string, data: Record<string, any>): string {
  if (src !== "smtp") throw new Error(`unknown render-fn source: ${src}`);
  const constructs = (data.records ?? []).map((r: any) => {
    const { name, priority, record, type, value } = r;
    let block: Record<string, any> = {
      zone_id: "${data.cloudflare_zone.domain.id}",
      name,
      ttl: "1",
      type,
      proxied: false,
    };
    if (type === "TXT") block = { ...block, content: `"${value}"` };
    if (type === "MX") block = { ...block, priority, content: value };
    return construct(new Construct(
      "resource",
      "cloudflare_dns_record",
      addSuffix("io.github.bigconig-ai.once.tools/smtp-dns", `-${record}-${type}`),
      block,
    ));
  });
  const merged = constructs.length ? sortNestedMap(deepMerge(...constructs)) : {};
  return cljJson(merged);
}

function ansibleDataFn(data: Record<string, any>): Record<string, any> {
  const sudoer = data.sudoer ?? "root";
  const hosts = [data.ip ?? "64.227.72.100"];
  return { ...data, sudoer, hosts, users: [] };
}

function inventory(data: Record<string, any>): string {
  const { sudoer, hosts, users } = data;
  const liveUsers: any[] = (users ?? [])
    .filter((u: any) => !u.remove)
    .flatMap((u: any) => hosts.map((host: string) => ({ ...u, host })));
  const admins: any[] = [{ ansible_user: sudoer }].flatMap((a) =>
    hosts.map((host: string) => ({ ...a, host, name: sudoer })),
  );
  const usersHosts: Record<string, any> = {};
  for (const u of liveUsers) {
    usersHosts[`${u.name}@${u.host}`] = {
      ansible_host: u.host,
      ansible_user: u.name,
      uid: u.uid,
    };
  }
  const adminsHosts: Record<string, any> = {};
  for (const a of admins) {
    adminsHosts[`root@${a.host}`] = {
      ansible_host: a.host,
      ansible_user: a.name,
    };
  }
  return cljJson({ all: { children: { admin: { hosts: adminsHosts }, users: { hosts: usersHosts } } } });
}

function yamlScalar(v: any): string {
  if (v === null || v === undefined) return "null";
  if (typeof v === "boolean") return v ? "true" : "false";
  return String(v);
}

function yamlLines(value: any, indent: string): string[] {
  if (Array.isArray(value)) {
    const lines: string[] = [];
    for (const item of value) {
      if (item !== null && typeof item === "object" && !Array.isArray(item)) {
        const sub = yamlLines(item, `${indent}  `);
        if (sub.length === 0) {
          lines.push(`${indent}- {}`);
          continue;
        }
        lines.push(`${indent}- ${sub[0].slice(indent.length + 2)}`);
        for (let i = 1; i < sub.length; i++) lines.push(sub[i]);
      } else {
        lines.push(`${indent}- ${yamlScalar(item)}`);
      }
    }
    return lines;
  }
  if (value !== null && typeof value === "object") {
    const lines: string[] = [];
    for (const [k, v] of Object.entries(value)) {
      if (Array.isArray(v)) {
        lines.push(`${indent}${k}:`);
        lines.push(...yamlLines(v, indent));
      } else if (v !== null && typeof v === "object") {
        lines.push(`${indent}${k}:`);
        lines.push(...yamlLines(v, `${indent}  `));
      } else {
        lines.push(`${indent}${k}: ${yamlScalar(v)}`);
      }
    }
    return lines;
  }
  return [`${indent}${yamlScalar(value)}`];
}

function toYaml(value: any): string {
  return `${yamlLines(value, "").join("\n")}\n`;
}

function ansibleOnce(data: Record<string, any>): string {
  const { once, domain } = data;
  const smtp: Record<string, any> = {};
  for (const k of ["smtp_server", "smtp_port", "smtp_username", "smtp_password"]) {
    if (k in data) smtp[k] = data[k];
  }
  smtp.smtp_from = `Info <info@notifications.${domain}>`;
  return toYaml([
    {
      name: "Reconcile ONCE applications",
      become: true,
      once: {
        ...once,
        applications: (once?.applications ?? []).map((app: any) => ({ ...app, ...smtp })),
      },
    },
  ]);
}

/** Multi-target render function for the Ansible inventory and ONCE task file. */
export function render(target: string, data: Record<string, any>): string {
  if (target === "inventory") return inventory(data);
  if (target === "ansible-once") return ansibleOnce(data);
  throw new Error(`unknown render target: ${target}`);
}

export function tofu(sfns: StepFn[], opts: Opts): Opts {
  const providerCompute = providerParam(opts, "provider-compute", "hcloud");
  const prepared = prepare(
    {
      [WF_NAME]: TOFU,
      [RENDER_TEMPLATES]: [
        {
          template: keywordToPath(TOFU),
          overwrite: true,
          "provider-compute": providerCompute,
          "compute-prevent-destroy": true,
          transform: [[providerCompute, delimiters]],
        },
      ],
    },
    opts,
  );
  return runStepsWithPlugin(pluginStep, sfns, prepared);
}

export function tofuSmtp(sfns: StepFn[], opts: Opts): Opts {
  const providerSmtp = providerParam(opts, "provider-smtp", "resend");
  const prepared = prepare(
    {
      [WF_NAME]: TOFU_SMTP,
      [RENDER_TEMPLATES]: [
        {
          template: keywordToPath(TOFU_SMTP),
          overwrite: true,
          "data-fn": ipDataFn,
          "provider-smtp": providerSmtp,
          transform: [[providerSmtp, delimiters]],
        },
      ],
    },
    opts,
  );
  return runStepsWithPlugin(pluginStep, sfns, prepared);
}

export function tofuDns(sfns: StepFn[], opts: Opts): Opts {
  const providerDns = providerParam(opts, "provider-dns", "cloudflare");
  const prepared = prepare(
    {
      [WF_NAME]: TOFU_DNS,
      [RENDER_TEMPLATES]: [
        {
          template: keywordToPath(TOFU_DNS),
          overwrite: true,
          "data-fn": ipDataFn,
          "provider-dns": providerDns,
          transform: [
            [providerDns, delimiters],
            [renderFn, { smtp: "smtp.tf.json" }, delimiters],
          ],
        },
      ],
    },
    opts,
  );
  return runStepsWithPlugin(pluginStep, sfns, prepared);
}

export function tofuSmtpPost(sfns: StepFn[], opts: Opts): Opts {
  const providerSmtp = providerParam(opts, "provider-smtp", "resend");
  const prepared = prepare(
    {
      [WF_NAME]: TOFU_SMTP_POST,
      [RENDER_TEMPLATES]: [
        {
          template: keywordToPath(TOFU_SMTP_POST),
          overwrite: true,
          "provider-smtp": providerSmtp,
          transform: [[providerSmtp, delimiters]],
        },
      ],
    },
    opts,
  );
  return runStepsWithPlugin(pluginStep, sfns, prepared);
}

export function ansible(sfns: StepFn[], opts: Opts): Opts {
  const prepared = prepare(
    {
      [WF_NAME]: ANSIBLE,
      [RENDER_TEMPLATES]: [
        {
          template: keywordToPath(ANSIBLE),
          overwrite: true,
          "data-fn": ansibleDataFn,
          transform: [
            [".", delimiters],
            [render, { inventory: "inventory.json", "ansible-once": "once.yml" }, delimiters],
          ],
        },
      ],
    },
    opts,
  );
  return runSteps(sfns, prepared);
}

export function ansibleLocal(sfns: StepFn[], opts: Opts): Opts {
  const prepared = prepare(
    {
      [WF_NAME]: ANSIBLE_LOCAL,
      [RENDER_TEMPLATES]: [
        {
          template: keywordToPath(ANSIBLE_LOCAL),
          overwrite: true,
          transform: [["."]],
        },
      ],
    },
    opts,
  );
  return runSteps(sfns, prepared);
}

function toolStar(
  fn: (sfns: StepFn[], opts: Opts) => Opts,
): (args: string | string[], opts?: Opts) => Opts {
  return (args, opts = {}) => fn(stepFns, { ...parseArgs(args), "big-config/env": "shell", ...opts });
}

export const tofuStar = toolStar(tofu);
export const tofuSmtpStar = toolStar(tofuSmtp);
export const tofuDnsStar = toolStar(tofuDns);
export const tofuSmtpPostStar = toolStar(tofuSmtpPost);
export const ansibleStar = toolStar(ansible);
export const ansibleLocalStar = toolStar(ansibleLocal);

export { runStepsWithPlugin };

File: src/once/tools.py
"""Tofu / Ansible tool workflows."""
from __future__ import annotations

from typing import Any, Callable, Iterable

from big_config import ERR, EXIT
from big_config import pluggable
from big_config import render as bc_render
from big_config import workflow as bc_workflow
from big_config.core import Opts, StepFn
from big_config.step_fns import exit_step_fn, print_error_step_fn
from big_config.utils import deep_merge, keyword_to_path, sort_nested_map
from big_tofu.core import Construct, add_suffix, construct

from .interop import PARAMS

END = "big-config.workflow/end"

step_fns: list[StepFn] = [bc_workflow.print_step_fn, exit_step_fn(END), print_error_step_fn(END)]

delimiters = {"tag-open": "<", "tag-close": ">", "filter-open": "{", "filter-close": "}"}

TOFU = "io.github.amiorin.once.tools/tofu"
TOFU_SMTP = "io.github.amiorin.once.tools/tofu-smtp"
TOFU_DNS = "io.github.amiorin.once.tools/tofu-dns"
TOFU_SMTP_POST = "io.github.amiorin.once.tools/tofu-smtp-post"
ANSIBLE_LOCAL = "io.github.amiorin.once.tools/ansible-local"
ANSIBLE = "io.github.amiorin.once.tools/ansible"

plugin_step = "io.github.amiorin.once.tools/render-tofu-backend"


def run_steps_with_plugin(plugin: str, sfns: list[StepFn], opts: Opts) -> Opts:
    steps: list[str] = []
    for step in opts.get(bc_workflow.STEPS) or []:
        if step == "render":
            steps.extend([step, plugin])
        else:
            steps.append(step)
    return bc_workflow.run_steps(sfns, {**opts, bc_workflow.STEPS: steps})


def _provider_param(opts: Opts, key: str, default: Any) -> Any:
    return (opts.get(PARAMS) or {}).get(key, default)


def _render_tofu_backend(_f, _step: str, sfns: list[StepFn], opts: Opts) -> Opts:
    provider_backend = _provider_param(opts, "provider-backend", "s3")
    prepare_keys = [
        bc_workflow.NAME,
        bc_workflow.PATH_FN,
        bc_workflow.PREFIX,
        bc_workflow.OBJECT_FN,
        bc_workflow.OBJECT_PREFIX,
        bc_workflow.PARAMS,
    ]
    overrides = {k: opts[k] for k in prepare_keys if k in opts}
    prepared = bc_workflow.prepare(
        {
            bc_workflow.NAME: opts.get(bc_workflow.NAME),
            bc_render.TEMPLATES: [
                {
                    "template": keyword_to_path("io.github.amiorin.once.tools/tofu-backend"),
                    "overwrite": True,
                    "provider-backend": provider_backend,
                    "transform": [[provider_backend, delimiters]],
                }
            ],
        },
        overrides,
    )
    plugin_opts = bc_render.templates(sfns, prepared)
    return {
        **opts,
        EXIT: plugin_opts.get(EXIT),
        ERR: plugin_opts.get(ERR),
        plugin_step: [*(opts.get(plugin_step) or []), plugin_opts],
    }


pluggable.defmethod(plugin_step, _render_tofu_backend)


def _ip_data_fn(data: dict[str, Any], _opts: Opts | None = None) -> dict[str, Any]:
    return {**data, "ip": data.get("ip") or "192.168.0.1"}


def _json_string(value: str) -> str:
    import json

    return json.dumps(value, ensure_ascii=False)


def clj_json(value: Any, indent: int = 0) -> str:
    """Pretty JSON with Cheshire's spacing for the rendered fixtures."""
    sp = " " * indent
    child = " " * (indent + 2)
    if isinstance(value, dict):
        if not value:
            return "{ }"
        items = list(value.items())
        lines = ["{"]
        for i, (k, v) in enumerate(items):
            comma = "," if i < len(items) - 1 else ""
            lines.append(f"{child}{_json_string(str(k))} : {clj_json(v, indent + 2)}{comma}")
        lines.append(f"{sp}}}")
        return "\n".join(lines)
    if isinstance(value, list):
        if not value:
            return "[ ]"
        lines = ["["]
        for i, v in enumerate(value):
            comma = "," if i < len(value) - 1 else ""
            lines.append(f"{child}{clj_json(v, indent + 2)}{comma}")
        lines.append(f"{sp}]")
        return "\n".join(lines)
    if isinstance(value, str):
        return _json_string(value)
    if value is True:
        return "true"
    if value is False:
        return "false"
    if value is None:
        return "null"
    return str(value)


def render_fn(src: str, data: dict[str, Any]) -> str:
    """Build Cloudflare DNS record JSON from SMTP records."""
    if src != "smtp":
        raise ValueError(f"unknown render-fn source: {src}")
    constructs: list[dict[str, Any]] = []
    for r in data.get("records") or []:
        name = r.get("name")
        priority = r.get("priority")
        record = r.get("record")
        type_ = r.get("type")
        value = r.get("value")
        block: dict[str, Any] = {
            "zone_id": "${data.cloudflare_zone.domain.id}",
            "name": name,
            "ttl": "1",
            "type": type_,
            "proxied": False,
        }
        if type_ == "TXT":
            block = {**block, "content": f'"{value}"'}
        if type_ == "MX":
            block = {**block, "priority": priority, "content": value}
        constructs.append(
            construct(
                Construct(
                    "resource",
                    "cloudflare_dns_record",
                    add_suffix("io.github.amiorin.once.tools/smtp-dns", f"-{record}-{type_}"),
                    block,
                )
            )
        )
    merged = sort_nested_map(deep_merge(*constructs)) if constructs else {}
    return clj_json(merged)


def ansible_data_fn(data: dict[str, Any], _opts: Opts | None = None) -> dict[str, Any]:
    sudoer = data.get("sudoer") or "root"
    hosts = [data.get("ip") or "64.227.72.100"]
    return {**data, "sudoer": sudoer, "hosts": hosts, "users": []}


def inventory(data: dict[str, Any]) -> str:
    sudoer = data.get("sudoer")
    hosts = data.get("hosts") or []
    users = data.get("users") or []
    live_users = [{**u, "host": host} for u in users if not u.get("remove") for host in hosts]
    admins = [{**a, "host": host, "name": sudoer} for a in [{"ansible_user": sudoer}] for host in hosts]

    users_hosts = {
        f"{u.get('name')}@{u.get('host')}": {
            "ansible_host": u.get("host"),
            "ansible_user": u.get("name"),
            "uid": u.get("uid"),
        }
        for u in live_users
    }
    admins_hosts = {
        f"root@{a.get('host')}": {"ansible_host": a.get("host"), "ansible_user": a.get("name")}
        for a in admins
    }
    inv = {"all": {"children": {"admin": {"hosts": admins_hosts}, "users": {"hosts": users_hosts}}}}
    return clj_json(inv)


def _yaml_scalar(v: Any) -> str:
    if v is None:
        return "null"
    if isinstance(v, bool):
        return "true" if v else "false"
    return str(v)


def _yaml_lines(value: Any, indent: str = "") -> list[str]:
    if isinstance(value, list):
        lines: list[str] = []
        for item in value:
            if isinstance(item, dict):
                sub = _yaml_lines(item, indent + "  ")
                if not sub:
                    lines.append(f"{indent}- {{}}")
                    continue
                lines.append(f"{indent}- {sub[0][len(indent) + 2:]}")
                lines.extend(sub[1:])
            else:
                lines.append(f"{indent}- {_yaml_scalar(item)}")
        return lines
    if isinstance(value, dict):
        lines = []
        for k, v in value.items():
            if isinstance(v, list):
                lines.append(f"{indent}{k}:")
                lines.extend(_yaml_lines(v, indent))
            elif isinstance(v, dict):
                lines.append(f"{indent}{k}:")
                lines.extend(_yaml_lines(v, indent + "  "))
            else:
                lines.append(f"{indent}{k}: {_yaml_scalar(v)}")
        return lines
    return [f"{indent}{_yaml_scalar(value)}"]


def _to_yaml(value: Any) -> str:
    return "\n".join(_yaml_lines(value, "")) + "\n"


def ansible_once(data: dict[str, Any]) -> str:
    once = data.get("once") or {}
    domain = data.get("domain")
    smtp = {k: data[k] for k in ["smtp_server", "smtp_port", "smtp_username", "smtp_password"] if k in data}
    smtp["smtp_from"] = f"Info <info@notifications.{domain}>"
    tasks = [
        {
            "name": "Reconcile ONCE applications",
            "become": True,
            "once": {
                **once,
                "applications": [{**app, **smtp} for app in (once.get("applications") or [])],
            },
        }
    ]
    return _to_yaml(tasks)


def render(target: str, data: dict[str, Any]) -> str:
    if target == "inventory":
        return inventory(data)
    if target == "ansible-once":
        return ansible_once(data)
    raise ValueError(f"unknown render target: {target}")


def tofu(sfns: list[StepFn], opts: Opts) -> Opts:
    provider_compute = _provider_param(opts, "provider-compute", "hcloud")
    prepared = bc_workflow.prepare(
        {
            bc_workflow.NAME: TOFU,
            bc_render.TEMPLATES: [
                {
                    "template": keyword_to_path(TOFU),
                    "overwrite": True,
                    "provider-compute": provider_compute,
                    "compute-prevent-destroy": True,
                    "transform": [[provider_compute, delimiters]],
                }
            ],
        },
        opts,
    )
    return run_steps_with_plugin(plugin_step, sfns, prepared)


def tofu_smtp(sfns: list[StepFn], opts: Opts) -> Opts:
    provider_smtp = _provider_param(opts, "provider-smtp", "resend")
    prepared = bc_workflow.prepare(
        {
            bc_workflow.NAME: TOFU_SMTP,
            bc_render.TEMPLATES: [
                {
                    "template": keyword_to_path(TOFU_SMTP),
                    "overwrite": True,
                    "data-fn": _ip_data_fn,
                    "provider-smtp": provider_smtp,
                    "transform": [[provider_smtp, delimiters]],
                }
            ],
        },
        opts,
    )
    return run_steps_with_plugin(plugin_step, sfns, prepared)


def tofu_dns(sfns: list[StepFn], opts: Opts) -> Opts:
    provider_dns = _provider_param(opts, "provider-dns", "cloudflare")
    prepared = bc_workflow.prepare(
        {
            bc_workflow.NAME: TOFU_DNS,
            bc_render.TEMPLATES: [
                {
                    "template": keyword_to_path(TOFU_DNS),
                    "overwrite": True,
                    "data-fn": _ip_data_fn,
                    "provider-dns": provider_dns,
                    "transform": [
                        [provider_dns, delimiters],
                        [render_fn, {"smtp": "smtp.tf.json"}, delimiters],
                    ],
                }
            ],
        },
        opts,
    )
    return run_steps_with_plugin(plugin_step, sfns, prepared)


def tofu_smtp_post(sfns: list[StepFn], opts: Opts) -> Opts:
    provider_smtp = _provider_param(opts, "provider-smtp", "resend")
    prepared = bc_workflow.prepare(
        {
            bc_workflow.NAME: TOFU_SMTP_POST,
            bc_render.TEMPLATES: [
                {
                    "template": keyword_to_path(TOFU_SMTP_POST),
                    "overwrite": True,
                    "provider-smtp": provider_smtp,
                    "transform": [[provider_smtp, delimiters]],
                }
            ],
        },
        opts,
    )
    return run_steps_with_plugin(plugin_step, sfns, prepared)


def ansible(sfns: list[StepFn], opts: Opts) -> Opts:
    prepared = bc_workflow.prepare(
        {
            bc_workflow.NAME: ANSIBLE,
            bc_render.TEMPLATES: [
                {
                    "template": keyword_to_path(ANSIBLE),
                    "overwrite": True,
                    "data-fn": ansible_data_fn,
                    "transform": [
                        [".", delimiters],
                        [render, {"inventory": "inventory.json", "ansible-once": "once.yml"}, delimiters],
                    ],
                }
            ],
        },
        opts,
    )
    return bc_workflow.run_steps(sfns, prepared)


def ansible_local(sfns: list[StepFn], opts: Opts) -> Opts:
    prepared = bc_workflow.prepare(
        {
            bc_workflow.NAME: ANSIBLE_LOCAL,
            bc_render.TEMPLATES: [
                {
                    "template": keyword_to_path(ANSIBLE_LOCAL),
                    "overwrite": True,
                    "transform": [["."]],
                }
            ],
        },
        opts,
    )
    return bc_workflow.run_steps(sfns, prepared)


def tool_star(fn: Callable[[list[StepFn], Opts], Opts]):
    def star(args: str | list[str], opts: Opts | None = None) -> Opts:
        parsed = bc_workflow.parse_args(args)
        return fn(step_fns, {**parsed, "big-config/env": "shell", **(opts or {})})

    return star


tofu_star = tool_star(tofu)
tofu_smtp_star = tool_star(tofu_smtp)
tofu_dns_star = tool_star(tofu_dns)
tofu_smtp_post_star = tool_star(tofu_smtp_post)
ansible_star = tool_star(ansible)
ansible_local_star = tool_star(ansible_local)

# TypeScript-style aliases.
runStepsWithPlugin = run_steps_with_plugin
pluginStep = plugin_step
renderFn = render_fn
tofuSmtp = tofu_smtp
tofuDns = tofu_dns
tofuSmtpPost = tofu_smtp_post
ansibleLocal = ansible_local
tofuStar = tofu_star
tofuSmtpStar = tofu_smtp_star
tofuDnsStar = tofu_dns_star
tofuSmtpPostStar = tofu_smtp_post_star
ansibleStar = ansible_star
ansibleLocalStar = ansible_local_star
